graph/db/sql_store.go

```diff
@@ -2230,6 +2238,211 @@ func (s *SQLStore) PruneGraphNodes() ([]route.Vertex, error) {
+		closedChans []*models.ChannelEdgeInfo
+		prunedNodes []route.Vertex
+	)
+	err := s.db.ExecTx(ctx, sqldb.WriteTxOpt(), func(db SQLQueries) error {
+		for _, outpoint := range spentOutputs {
+			// TODO(elle): potentially optimize this by using
+			//  sqlc.slice() once that works for both SQLite and
+			//  Postgres.
+			//
+			// NOTE: this fetches channels for all protocol
+			// versions.
+			row, err := db.GetChannelByOutpoint(
+				ctx, outpoint.String(),
+			)
+			if errors.Is(err, sql.ErrNoRows) {
+				continue
+			} else if err != nil {
+				return fmt.Errorf("unable to fetch channel: %w",
+					err)
+			}
+
+			node1, node2, err := buildNodeVertices(
+				row.Node1Pubkey, row.Node2Pubkey,
+			)
+			if err != nil {
+				return err
+			}
+
+			info, err := getAndBuildEdgeInfo(
+				ctx, db, s.cfg.ChainHash, row.Channel.ID,
+				row.Channel, node1, node2,
+			)
+			if err != nil {
+				return err
+			}
+
+			err = db.DeleteChannel(ctx, row.Channel.ID)
+			if err != nil {
+				return fmt.Errorf("unable to delete "+
+					"channel: %w", err)
+			}
+
+			closedChans = append(closedChans, info)
+		}
+
+		err := db.UpsertPruneLogEntry(
+			ctx, sqlc.UpsertPruneLogEntryParams{
+				BlockHash:   blockHash[:],
+				BlockHeight: int64(blockHeight),
+			},
+		)
+		if err != nil {
+			return fmt.Errorf("unable to insert prune log "+
+				"entry: %w", err)
+		}
+
+		// Now that we've pruned some channels, we'll also prune any
+		// nodes that no longer have any channels.
+		prunedNodes, err = s.pruneGraphNodes(ctx, db)
+		if err != nil {
+			return fmt.Errorf("unable to prune graph nodes: %w",
+				err)
+		}
+
+		return nil
+	}, func() {
+		prunedNodes = nil
+		closedChans = nil
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to prune graph: %w", err)
+	}
+
+	for _, channel := range closedChans {
+		s.rejectCache.remove(channel.ChannelID)
+		s.chanCache.remove(channel.ChannelID)
+	}
+
+	return closedChans, prunedNodes, nil
+}
+
+// ChannelView returns the verifiable edge information for each active channel
+// within the known channel graph. The set of UTXOs (along with their scripts)
+// returned are the ones that need to be watched on chain to detect channel
+// closes on the resident blockchain.
+//
+// NOTE: part of the V1Store interface.
+func (s *SQLStore) ChannelView() ([]EdgePoint, error) {
+	var (
+		ctx        = context.TODO()
+		edgePoints []EdgePoint
+	)
+
+	handleChannel := func(db SQLQueries, channel sqlc.Channel) error {
+		pkScript, err := genMultiSigP2WSH(
+			channel.BitcoinKey1, channel.BitcoinKey2,
+		)
+		if err != nil {
+			return err
+		}
+
+		op, err := wire.NewOutPointFromString(channel.Outpoint)
+		if err != nil {
+			return err
+		}
+
+		edgePoints = append(edgePoints, EdgePoint{
+			FundingPkScript: pkScript,
+			OutPoint:        *op,
+		})
+
+		return nil
+	}
+
+	err := s.db.ExecTx(ctx, sqldb.ReadTxOpt(), func(db SQLQueries) error {
+		lastID := int64(-1)
+		for {
+			rows, err := db.ListChannelsPaginated(
```

Comment:
nit: since we iterate through all of the channels, it'd be more efficient to only fetch the necessary data (the keys and the outpoint).